### ISSUE TITLE
Versioned::allVersions() somewhat broken

### ISSUE THREAD

**Issue by dhensby** [silverstripe/silverstripe-framework#7663]
https://github.com/silverstripe/silverstripe-framework/blob/3.5/model/Versioned.php#L1038-L1090

`$baseTable` isn't defined if line 1059 doesn't ever evaluate to true

`$having` parameter is useless as there is no `having` method on `DataList`. If `$having` is supplied expect an error to be triggered

# PRs

 - [x] https://github.com/silverstripe/silverstripe-versioned/pull/81
 - [x] https://github.com/silverstripe/silverstripe-cms/pull/2065
 - [x] https://github.com/silverstripe/silverstripe-framework/pull/7737

**Comment by tractorcow** [comment]
Why can't we just use `$this->owner->baseTable()`? :P

We also have this on 4.0... let's just make it an error and completely remove this arg (method?) in master.

```php
if ($having) {
            // @todo - This method doesn't exist on DataList
            $list->having($having);
        }
```

This looks like one of the few versioned methods I didn't completely re-write... I'd have done it differently if so. :P

**Comment by tractorcow** [comment]
I'm going to be super lazy and guess this might work without testing it.

```
/**
 * Filterable datalist that returns all versions
 *
 * @return DataList
 */
public function allVersions()
{
	return DataObject::get($this->owner->baseClass())
		->distinct(false)
		->setDataQueryParam('Versioned.mode', 'all_versions');
}
```

I would hope the ORM / augmentSQL() would be smart enough to return the correct list for me.

**Comment by dhensby** [comment]
Can't we just use `Versioned::get_all_versions()` (which is basically the code you've written out)

https://github.com/silverstripe/silverstripe-versioned/blob/1.0/src/Versioned.php#L2516-L2531

**Comment by tractorcow** [comment]
Oh right, yeah that is pretty much the same method isn't it?

... so why does allVersions() do what it does...

**Comment by tractorcow** [comment]
Oh right, because it returns a list of `Versioned_Version` instead, not the object itself.

solution: Drop Versioned_Version :D Looking at that class it seems super redundant.

**Comment by dhensby** [comment]
agreed

**Comment by tractorcow** [comment]
PR for 5.0 at https://github.com/silverstripe/silverstripe-versioned/pull/81. Linkes to a separate framework changelog change and CMS regression fix for master / 5.0.

**Comment by robbieaverill** [comment]
All pull requests are merged, feel free to reopen if there is still work to be done
